File: 2020-08-05/neo.ts
function searchRange(nums: number[], target: number): number[] {
	const from = lowerBound(nums, target);
	if (from === nums.length || nums[from] !== target) {
		return [-1, -1];
	}
	const to = lowerBound(nums, target + 1, from);
	return [from, to - 1];
}

function lowerBound(nums: number[], target: number, from: number = 0, to: number = nums.length): number {
	while (from < to) {
		let mid = (from + to) >> 1;
		if (target <= nums[mid]) {
			to = mid;
		} else {
			from = mid + 1;
		}
	}
	return to;
}

File: 2020-08-12/neo.kt
class Solution {
    fun searchRange(nums: IntArray, target: Int): IntArray {
        val from = nums.lowerBound(target)
        val to = nums.lowerBound(target + 1, from, nums.size)
        return if (from == to) intArrayOf(-1, -1)
        else intArrayOf(from, to - 1)
    }
}

fun IntArray.lowerBound(target: Int, fromIndex: Int = 0, toIndex: Int = this.size): Int {
    var l = fromIndex
    var r = toIndex
    while (l < r) {
        val mid = (l + r) / 2
        if (target <= this[mid]) {
            r = mid
        } else {
            l = mid + 1
        }
    }
    return r
}
